pkger/service_logging.go

```diff
@@ -0,0 +1,92 @@
+package pkger
+
+import (
+	"context"
+	"time"
+
+	"github.com/influxdata/influxdb"
+	"go.uber.org/zap"
+)
+
+type loggingMW struct {
+	logger *zap.Logger
+	next   SVC
+}
+
+// MWLogging adds logging functionality for the service.
+func MWLogging(log *zap.Logger) SVCMiddleware {
+	return func(svc SVC) SVC {
+		return &loggingMW{
+			logger: log,
+			next:   svc,
+		}
+	}
+}
+
+var _ SVC = (*loggingMW)(nil)
+
+func (s *loggingMW) CreatePkg(ctx context.Context, setters ...CreatePkgSetFn) (pkg *Pkg, err error) {
+	defer func(start time.Time) {
+		dur := zap.Duration("took", time.Since(start))
+		if err != nil {
+			s.logger.Error("failed to create pkg", zap.Error(err), dur)
+			return
+		}
+		fields := s.summaryLogFields(pkg.Summary())
+		s.logger.Info("pkg create", append(fields, dur)...)
+	}(time.Now())
+	return s.next.CreatePkg(ctx, setters...)
+}
+
+func (s *loggingMW) DryRun(ctx context.Context, orgID, userID influxdb.ID, pkg *Pkg) (sum Summary, diff Diff, err error) {
+	defer func(start time.Time) {
+		dur := zap.Duration("took", time.Since(start))
+		if err != nil {
```

Comment:
not sure this will work, have you test it out? Usually, i would pass an *error pointer as a parameter of your defer func

Reply:
this works b/c the enclosure isn't forcing the assignment. When you pass an error as a func arg, you'll need to that as `*error`

Reply:
nice